src/jit/importer.cpp

```diff
@@ -4269,6 +4283,13 @@ NamedIntrinsic Compiler::lookupNamedIntrinsic(CORINFO_METHOD_HANDLE method)
             result = NI_System_Collections_Generic_EqualityComparer_get_Default;
         }
     }
+    else if (strcmp(namespaceName, "System.Runtime.CompilerServices") == 0)
+    {
```

Comment:
Why do you need all this - why is not `object.GetRawData()` + `Unsafe.Add` enough?

Reply:
It was resulting in weird unoptimized codegen. (The Memory.Span property getter today shows some of this weird codegen since it uses a very similar trick to what you propose.) This seemed like the safest / most straightforward way to address it.

Reply:
My usual question for this is how hard is it to just fix it in the JIT. You can always introduce special intrinsics, etc., but it does not scale.

Reply:
Somebody more familiar with JIT would have to chime in, but a cursory inspection makes it seem like modifying the optimizer would be more work than adding this intrinsic. It seems JIT doesn't quite know how to optimize the two-step process of getting a field ref from an object (convert GCREF to BYREF), then further manipulating the BYREF by performing an addition operation. I suspect that there was simply never a scenario for this previously so it never made sense to invest in this seemingly bizarre optimization.

Reply:
Are you talking about generated code such as:
```
       3912                 cmp      dword ptr [rdx], edx
       4883C208             add      rdx, 8
       8B4204               mov      eax, dword ptr [rdx+4]
```
where the null check could be omitted and the addition could be folded into the address mode?

Reply:
Yes. As this PR currently stands I've added a grand total of 10 lines (+ comments and whitespace) to the JIT to enable this. It seemed like the most straightforward, lowest churn, least risky way to do it. If there's a way to modify the optimizer to fold the 3-instruction tuple ldflda + add (const) + add (variable) into a single node, I'd happily take that and drop the custom intrinsic. Per my other comment above, it seems the JIT was never set up to optimize this condition, because honestly when would you ever see ldflda + add + add in the IL stream?

Reply:
> I've added a grand total of 10 lines (+ comments and whitespace) to the JIT to enable this

The brand new tree shapes produced by the importer tend to have unexpected interations with the rest of the optimizations. For example, when we have added the byref intrinsics for Span, there was a long bug tail from iterations with the optimizer. 

> `ldflda + add (const) + add (variable)`

What is the problem with this pattern? E.g. I have tried:

```
ldarg.0
ldflda     uint8 F::myField
ldc.i4.5
add
ldarg.1
add
ldind.u1
ret
```

and it produced:

```
cmp      dword ptr [rcx], ecx
movsxd   rax, edx
movzx    rax, byte  ptr [rcx+rax+5]
ret
```

It is about as good code as it can be.

Reply:
In your case, it looks like `rcx` is already a byref to `F::myField`, not a gcref to `F`. (Is _F_ a struct passed as byref in your scenario?)

Take a look at the helper routine we use in `Memory<T>.get_Span`:

https://github.com/dotnet/coreclr/blob/aea7846fc71591739e47c65c0632007bff1cc4a4/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs#L167-L203

This results in the following codegen today:

```asm
; assume rcx := 'obj' and has already been validated as non-null
lea temp64, [rcx + 08h]  ; ref to obj._firstField
mov temp64, qword ptr [temp64 - 08h]  ; deref field at offset 0 in 'obj'
cmp 0, dword ptr [temp64]  ; compare *(int*)temp64 < 0
```

With this PR as proposed (or a JIT optimization that can smash ldflda + add together), the codegen is simplified:

```asm
; assume rcx := 'obj' and has already been validated as non-null
mov temp64, qword ptr [rcx]  ; deref field at offset 0 in 'obj'
cmp 0, dword ptr [temp64]  ; compare *(int*)temp64 < 0
```

Properly optimizing this in the JIT would also allow us to write the `FieldAccessor` code as follows:

```cs
nuint thisFieldOffset = _fieldOffset;
return ref Unsafe.As<byte, T>(
    ref Unsafe.AddByteOffset(
        ref Unsafe.AddByteOffset(ref obj.GetRawData(), -IntPtr.Size),
        thisFieldOffset));
```

Resulting in this codegen:

```
mov rax, qword ptr [this._fieldOffset]
cmp edx, qword ptr [rdx]  ; null check on obj (could be elided)
lea rax, [rdx + rax]  ; note optimized 2-argument lea form, not slow 3-argument form
```

Reply:
> ObjectHasComponentSize

I am not really happy that the current implementation blurs the boundaries between objectrefs and byrefs. E.g. it would prevent us from experimenting with removing syncblocks. The proper implementation of this method would avoid ever creating a byref pointing to offset 0.

Reply:
I feel you that it's a bit of a hack and has definitely got some code smell to it. And I'd prefer a much cleaner way of doing it. :)

But before it was checked in there were discussions with people familiar with the runtime, object layout, the JIT, and the GC. And it was determined that accessing fields at offset 0 (_pMethodTable_) would be safe.

You're right that accessing fields at truly negative offsets would be dangerous, and it would interfere both with the GC's ability to ref track and the runtime team's ability to experiment with things like sync block removal. But nobody is proposing accessing fields at negative offsets, let alone optimizing it.

Reply:
BTW I'm experimenting with whether it'd be possible to do this using IL replacement rather than GenTree replacement. Would that alleviate some of the concerns people have?

Reply:
Why can't you just do `=> ref Unsafe.As<byte, TField>(ref Unsafe.AddByteOffset(ref RawData.GetRawData(t), _offset));`? The codegen for that may have one extra instruction than the ideal codegen; but I do not think that it warrants workarounds like IL replacement.

> it resulted in incorrect codegen

Yep, `byref` arithmetic is very slippery slope. Many opportunities for hard to trace down GC holes.

Reply:
> The codegen for that may have one extra instruction than the ideal codegen; but I do not think that it warrants workarounds like IL replacement.

Sure, I'll cede that point. Bonus: using the existing code allows the _obj_ null check to avoid a register spill. :)